### PR TITLE
Refactor daily course enrollment

### PR DESCRIPTION
**Pull request by mulby**
@brianhw as discussed.

Note the introduction of HiveQueryToMysqlTask.

Still needs a bunch of docs and testing. A quick validation run indicates I didn't actually change the output at all.

### CODE REVIEW

File: edx/analytics/tasks/enrollments.py [hunk 1]
```diff
@@ -412,63 +433,27 @@ def insert_query(self):
                 ce.date,
                 ce.course_id,
                 IF(p.gender != '', p.gender, NULL)
-            """
+        """
 
     @property
-    def table_name(self):
+    def table(self):
         return 'course_enrollment_gender'
```
Comment (brianhw) [inline edx/analytics/tasks/enrollments.py]: update table name (_daily).

File: edx/analytics/tasks/enrollments.py [hunk 2]
```diff
@@ -482,63 +467,27 @@ def insert_query(self):
                 ce.date,
                 ce.course_id,
                 p.year_of_birth
-            """
+        """
 
     @property
-    def table_name(self):
+    def table(self):
         return 'course_enrollment_birth_year'
```
Comment (brianhw) [inline edx/analytics/tasks/enrollments.py]: update table name (_daily).

File: edx/analytics/tasks/enrollments.py [hunk 3]
```diff
@@ -552,57 +501,21 @@ def insert_query(self):
                 ce.date,
                 ce.course_id,
                 IF(p.level_of_education != '', p.level_of_education, NULL)
-            """
+        """
 
     @property
-    def table_name(self):
+    def table(self):
         return 'course_enrollment_education_level'
```
Comment (brianhw) [inline edx/analytics/tasks/enrollments.py]: update table name (_daily).

File: edx/analytics/tasks/enrollments.py [hunk 4]
```diff
@@ -308,76 +297,108 @@ def columns(self):
...
-    def table_format(self):
-        return "ROW FORMAT DELIMITED FIELDS TERMINATED BY '\\t'"
+    def partition(self):
+        return HivePartition('dt', self.blacklist_date)  # pylint: disable=no-member
 
 
-class EnrollmentDemographicTask(CourseEnrollmentTableDownstreamMixin, ImportIntoHiveTableTask):
-    """Base class for demographic breakdowns of enrollments"""
+class EnrollmentDailyTask(CourseEnrollmentTableDownstreamMixin, HiveQueryToMysqlTask):
+    """A history of the number of students enrolled in each course at the end of each day"""
```
Comment (brianhw) [inline edx/analytics/tasks/enrollments.py]: A general nit across many of the new docstrings:  the summary line should end with a period.

File: edx/analytics/tasks/enrollments.py [hunk 5]
```diff
@@ -290,7 +279,7 @@ def requires(self):
         )
 
 
-class EnrollmentCourseBlacklistTable(WarehouseMixin, ImportIntoHiveTableTask):
+class EnrollmentCourseBlacklistTable(HiveTableTask):
```
Comment (brianhw) [inline edx/analytics/tasks/enrollments.py]: I initially found the lack of "Task" suffix on the classname to be jarring.  It really is a TableTask, as the base class suggests.  Maybe it's not ambiguous -- that a "Table" class implies that it's also a task and not just a class that deals with a table.  But this also carries over to the use of "required_tables()".

File: edx/analytics/tasks/enrollments.py [hunk 6]
```diff
@@ -308,76 +297,108 @@ def columns(self):
...
     @property
-    def table_name(self):
-        raise NotImplementedError
+    def partition(self):
+        return HivePartition('dt', self.interval.date_b.isoformat())  # pylint: disable=no-member
 
     @property
-    def columns(self):
-        raise NotImplementedError
+    def required_tables(self):
```
Comment (brianhw) [inline edx/analytics/tasks/enrollments.py]: Continuation from above -- this uses the same shortening of Table to mean TableTask.  But here it goes further, where it implies that the requirements will all be other Table(Task)s.  So far that assumption holds.  Are there likely cases where one would have a requirement that was a Task but not a Table(Task)?  If so, would they be included in required_tables(), or do they get defined as required() with a call to super()?

**Comment by brianhw** [comment]
Looks great!  Just a reservation about nomenclature (Table as a Task) that I need to sleep on, and some docstring nits.

**Comment by mulby** [comment]
@brianhw I have pushed an update to this patch after rebasing onto the tip of master and cleaned up things a bit.

The enrollment acceptance tests pass, still waiting on the rest of the suite to finish.

**Comment by mulby** [comment]
Acceptance tests passed:

http://jenkins-ci.analytics.edx.org/view/pull%20requests/job/edx-analytics-pipeline-acceptance-auto-pr/183/

**Comment by brianhw** [comment]
LGTM.  Needs a squash of history, then :+1:

**Comment by mulby** [comment]
@rocha can you take a look at this too? This introduces the "single task load from Hive to MySQL" we discussed in the other PR.

File: edx/analytics/tasks/enrollments.py (hunk 5 above)
Comment (rocha) [inline edx/analytics/tasks/enrollments.py]: +1

**Comment by rocha** [comment]
Love it! :+1:

**Comment by mulby** [comment]
Alright, I did a full run with this code and compared with prod (up to 12/3) and it matched 100%. I'll do the rename, run a final test and then merge.

**Comment by rocha** [comment]
:shipit:

**Pull request merged**